Darwin/StitchCore/StitchCore/Core/Auth/StitchAuthDelegate.swift

```diff
@@ -13,7 +15,171 @@ public protocol StitchAuthDelegate: class {
+     *
+     * - parameters:
+     *    - auth: The instance of `StitchAuth` where the user was linked.
+     *            It can be used to infer the current state of authentication.
+     *    - linkedUser: The user that was linked to a new identity.
+     */
+    func onUserLinked(auth: StitchAuth, linkedUser: StitchUser)
+
+    /**
+     * Called whenever a user is logged out. The user logged out is not
+     * necessarily the active user. If the user logged out was the active user,
+     * then `onActiveUserChanged` will be called after this method. If the user
+     * was an anonymous user, that user will also be removed and
+     * `onUserRemoved` will also be called.
+     *
+     * - parameters:
+     *    - auth: The instance of `StitchAuth` where the user was logged out.
+     *            It can be used to infer the current state of authentication.
+     *    - loggedOutUser: The user that was logged out.
+     */
+    func onUserLoggedOut(auth: StitchAuth, loggedOutUser: StitchUser)
+
+    /**
+     * Called whenever the active user changes. This may be due to a call to
+     * `StitchAuth.loginWithCredential`, `StitchAuth.switchToUserWithId`,
+     * `StitchAuth.logout`, `StitchAuth.logoutUserWithId`,
+     * `StitchAuth.removeUser`, or `StitchAuth.removeUserWithId`.
+     * This may also occur on a normal request if a user's session is invalidated
+     * and they are forced to log out.
+     *
+     * - parameters:
+     *    - auth: The instance of `StitchAuth` where the the active user changed.
+     *            It can be used to infer the current state of authentication.
+     *    - currentActiveUser: The active user after the change.
+     *    - previousActiveUser: The active user before the change.
+     */
+    func onActiveUserChanged(auth: StitchAuth,
+                             currentActiveUser: StitchUser?,
+                             previousActiveUser: StitchUser?)
+
+    /**
+     * Called whenever a user is removed from the list of users on the device.
+     *
+     * - parameters:
+     *    - auth: The instance of `StitchAuth` where the the user was removed.
+     *            It can be used to infer the current state of authentication.
+     *    - removedUser: The user that was removed.
+     */
+    func onUserRemoved(auth: StitchAuth, removedUser: StitchUser)
+
+    /**
+     * Called whenever this delegate is registered for the first time. This can
+     * be useful to infer the state of authentication, because any events that
+     * occurred before the delegate was registered will not be seen by the
+     * delegate.
+     *
+     * - parameters:
+     *    - auth: The instance of `StitchAuth` where the the delegate was registered.
+     *            It can be used to infer the current state of authentication.
+     */
+    func onDelegateRegistered(auth: StitchAuth)
+}
+
+// extensions that provide default implementations of the the auth delegate methods
```

Comment:
Im confused by the pattern that is going on here. I understand delegates, but I don't get what is going on below with the extension if you have a minute to explain it to me.

Reply:
When people implement `StitchAuthDelegate`, we don't want them to be required to implement every single possible method (i.e., not all users will need to implement somtehing for `onUserAdded` or `onUserRemoved`, some users might care only about `onActiveUserChanged`).

By adding these extensions to the protocol, classes that implement the protocol will have these implementations by default.

It's why in the StitchAuthDelegateIntTests, we're able to do something like this:

```swift
        class TestAuthDelegate: StitchAuthDelegate {
            let work = DispatchGroup()
            var numCalls = 0

            func onUserAdded(auth: StitchAuth, addedUser: StitchUser) {
                XCTAssertNotNil(addedUser.id)
                numCalls += 1
                work.leave()
            }
        }
```

See https://docs.swift.org/swift-book/LanguageGuide/Protocols.html#ID521 for more context.

Reply:
Yeah, I just throught that protocols are only method declarations